File: web/src/lib/conversation/session-key.ts
import { resolveAgentId } from "@/config/runtime-options";

const AGENT_SESSION_PREFIX = "agent";
const ROOM_SESSION_PREFIX = "room";
const ROOM_SHARED_SESSION_PREFIX = "room:group:";
const TOPIC_SEGMENT = "topic";
const ACCOUNT_SEGMENT = "acct";

export interface BuildSessionKeyOptions {
  channel: string;
  chat_type: string;
  ref: string;
  agent_id?: string | null;
  account_id?: string | null;
  thread_id?: string | null;
}

type SessionKeyKind = "agent" | "room" | "unknown";

export interface ParsedSessionKey {
  raw: string;
  kind: SessionKeyKind;
  is_structured: boolean;
  is_shared: boolean;
  agent_id: string | null;
  channel: string | null;
  chat_type: string | null;
  account_id: string | null;
  ref: string | null;
  thread_id: string | null;
  conversation_id: string | null;
}

interface AgentRefParts {
  account_id: string | null;
  error: string | null;
  ref_start: number;
}

interface ParsedAgentReference {
  account_id: string | null;
  ref: string | null;
  thread_id: string | null;
}

interface SessionKeyRule {
  parse: (sessionKey: string, isStructured: boolean) => ParsedSessionKey;
  prefix: string;
  validate: (sessionKey: string) => string | null;
}

function createParsedSessionKey(raw: string): ParsedSessionKey {
  return {
    raw,
    kind: "unknown",
    is_structured: false,
    is_shared: false,
    agent_id: null,
    channel: null,
    chat_type: null,
    account_id: null,
    ref: null,
    thread_id: null,
    conversation_id: null,
  };
}

function findTopicIndex(parts: string[], minIndex: number): number {
  return parts.findIndex(
    (part, index) => part === TOPIC_SEGMENT && index >= minIndex,
  );
}

function agentSessionKeyShapeError(): string {
  return "session_key must match agent:<agent_id>:<channel>:<chat_type>[:acct:<account_id>]:<ref>[:topic:<thread_id>]";
}

function splitAgentRefParts(parts: string[]): AgentRefParts {
  if (parts[4] !== ACCOUNT_SEGMENT) {
    return { account_id: null, error: null, ref_start: 4 };
  }
  if (parts.length < 7) {
    return {
      account_id: null,
      error: agentSessionKeyShapeError(),
      ref_start: 0,
    };
  }
  const accountId = parts[5]?.trim() ?? "";
  return accountId
    ? { account_id: accountId, error: null, ref_start: 6 }
    : {
        account_id: null,
        error: "session_key account_id is required after acct segment",
        ref_start: 0,
      };
}

function parseAgentReference(parts: string[]): ParsedAgentReference {
  const split = splitAgentRefParts(parts);
  if (split.error) {
    return { account_id: null, ref: null, thread_id: null };
  }

  // `:topic:` 是协议边界，ref 内部仍可包含冒号。
  const topicIndex = findTopicIndex(parts, split.ref_start);
  const hasTopic = topicIndex >= 0;
  const refEnd = hasTopic ? topicIndex : parts.length;
  return {
    account_id: split.account_id,
    ref: parts.slice(split.ref_start, refEnd).join(":") || null,
    thread_id: hasTopic
      ? parts.slice(topicIndex + 1).join(":") || null
      : null,
  };
}

function hasAgentSessionHeader(parts: string[]): boolean {
  return parts.length >= 5 && parts.slice(1, 4).every(Boolean);
}

function validateAgentReference(
  parts: string[],
  refStart: number,
): string | null {
  const topicIndex = findTopicIndex(parts, refStart);
  if (topicIndex < 0) {
    return parts.slice(refStart).join(":").trim()
      ? null
      : agentSessionKeyShapeError();
  }
  const ref = parts.slice(refStart, topicIndex).join(":").trim();
  const threadId = parts.slice(topicIndex + 1).join(":").trim();
  return ref && threadId ? null : agentSessionKeyShapeError();
}

function validateAgentSessionKey(sessionKey: string): string | null {
  const parts = sessionKey.split(":");
  if (!hasAgentSessionHeader(parts)) {
    return agentSessionKeyShapeError();
  }

  const split = splitAgentRefParts(parts);
  if (split.error) {
    return split.error;
  }
  return validateAgentReference(parts, split.ref_start);
}

function parseAgentSessionKey(
  sessionKey: string,
  isStructured: boolean,
): ParsedSessionKey {
  const parts = sessionKey.split(":");
  return {
    ...createParsedSessionKey(sessionKey),
    ...parseAgentReference(parts),
    kind: "agent",
    is_structured: isStructured,
    agent_id: resolveAgentId(parts[1]),
    channel: parts[2] || null,
    chat_type: parts[3] || "dm",
  };
}

function validateRoomSessionKey(sessionKey: string): string | null {
  const parts = sessionKey.split(":");
  const conversationId = parts.slice(2).join(":").trim();
  return parts.length >= 3 && parts[1] === "group" && conversationId
    ? null
    : "session_key must match room:group:<conversation_id>";
}

function parseRoomSessionKey(
  sessionKey: string,
  isStructured: boolean,
): ParsedSessionKey {
  const parts = sessionKey.split(":");
  const conversationId = parts.slice(2).join(":").trim() || null;
  return {
    ...createParsedSessionKey(sessionKey),
    kind: "room",
    is_structured: isStructured,
    is_shared: isStructured,
    chat_type: parts[1] || "group",
    ref: conversationId,
    conversation_id: conversationId,
  };
}

const SESSION_KEY_RULES: SessionKeyRule[] = [
  {
    parse: parseAgentSessionKey,
    prefix: `${AGENT_SESSION_PREFIX}:`,
    validate: validateAgentSessionKey,
  },
  {
    parse: parseRoomSessionKey,
    prefix: `${ROOM_SESSION_PREFIX}:`,
    validate: validateRoomSessionKey,
  },
];

function findSessionKeyRule(sessionKey: string): SessionKeyRule | undefined {
  return SESSION_KEY_RULES.find((rule) => sessionKey.startsWith(rule.prefix));
}

function getSessionKeyValidationError(
  sessionKey: string | null | undefined,
): string | null {
  const normalizedKey = (sessionKey ?? "").trim();
  if (!normalizedKey) {
    return "session_key is required";
  }
  const rule = findSessionKeyRule(normalizedKey);
  return rule
    ? rule.validate(normalizedKey)
    : "session_key must use structured gateway format";
}

/** 前后端必须共用协议构造入口，避免业务层手拼身份。 */
export function buildSessionKey({
  channel,
  chat_type: chatType,
  ref,
  agent_id: agentId,
  account_id: accountId,
  thread_id: threadId,
}: BuildSessionKeyOptions): string {
  const segments = [
    AGENT_SESSION_PREFIX,
    resolveAgentId(agentId),
    channel.trim(),
    chatType.trim(),
  ];
  const resolvedAccountId = accountId?.trim();
  if (resolvedAccountId) {
    segments.push(ACCOUNT_SEGMENT, resolvedAccountId);
  }
  segments.push(ref.trim());
  const resolvedThreadId = threadId?.trim();
  if (resolvedThreadId) {
    segments.push(TOPIC_SEGMENT, resolvedThreadId);
  }
  return segments.join(":");
}

export function buildRoomSharedSessionKey(conversationId: string): string {
  return `${ROOM_SHARED_SESSION_PREFIX}${conversationId}`;
}

export function buildRoomAgentSessionKey(
  conversationId: string,
  agentId: string,
  roomType: "dm" | "room" = "room",
): string {
  return buildSessionKey({
    channel: "ws",
    chat_type: roomType === "dm" ? "dm" : "group",
    ref: conversationId,
    agent_id: agentId,
  });
}

export function isStructuredSessionKey(sessionKey: string): boolean {
  return getSessionKeyValidationError(sessionKey) === null;
}

export function assertStructuredSessionKey(
  sessionKey: string | null | undefined,
): string {
  const errorMessage = getSessionKeyValidationError(sessionKey);
  if (errorMessage) {
    throw new Error(errorMessage);
  }
  return (sessionKey ?? "").trim();
}

export function parseSessionKey(
  sessionKey: string | null | undefined,
): ParsedSessionKey {
  const normalizedKey = (sessionKey ?? "").trim();
  const rule = findSessionKeyRule(normalizedKey);
  return rule
    ? rule.parse(normalizedKey, rule.validate(normalizedKey) === null)
    : createParsedSessionKey(normalizedKey);
}

export function getSessionKeyIdentity(
  sessionKey: string | null | undefined,
): string | null {
  const parsed = parseSessionKey(sessionKey);
  if (!parsed.raw) {
    return null;
  }

  // Room 身份只认 conversationId，避免别名演进影响前端比较。
  if (parsed.kind === "room" && parsed.conversation_id) {
    return `${ROOM_SESSION_PREFIX}:${parsed.conversation_id}`;
  }
  return parsed.raw;
}

export function areEquivalentSessionKeys(
  left: string | null | undefined,
  right: string | null | undefined,
): boolean {
  const leftIdentity = getSessionKeyIdentity(left);
  const rightIdentity = getSessionKeyIdentity(right);
  return Boolean(leftIdentity && rightIdentity && leftIdentity === rightIdentity);
}

File: internal/protocol/session_key.go
package protocol

import (
	"errors"
	"fmt"
	"strings"
)

const (
	// SessionChannelWebSocketSegment 表示 session_key 中的 WebSocket 通道段。
	SessionChannelWebSocketSegment = "ws"
	// SessionChannelDiscordSegment 表示 session_key 中的 Discord 通道段。
	SessionChannelDiscordSegment = "dg"
	// SessionChannelTelegramSegment 表示 session_key 中的 Telegram 通道段。
	SessionChannelTelegramSegment = "tg"
	// SessionChannelDingTalkSegment 表示 session_key 中的钉钉通道段。
	SessionChannelDingTalkSegment = "dt"
	// SessionChannelWeChatSegment 表示 session_key 中的微信通道段。
	SessionChannelWeChatSegment = "wx"
	// SessionChannelWeixinPersonalSegment 表示个人微信 iLink 通道段。
	SessionChannelWeixinPersonalSegment = "weixin-personal"
	// SessionChannelFeishuSegment 表示 session_key 中的飞书通道段。
	SessionChannelFeishuSegment = "fs"
	// SessionChannelInternalSegment 表示 session_key 中的内部通道段。
	SessionChannelInternalSegment = "internal"

	// SessionChannelWebSocket 表示持久化后的 WebSocket 通道类型。
	SessionChannelWebSocket = "websocket"
	// SessionChannelDiscord 表示持久化后的 Discord 通道类型。
	SessionChannelDiscord = "discord"
	// SessionChannelTelegram 表示持久化后的 Telegram 通道类型。
	SessionChannelTelegram = "telegram"
	// SessionChannelDingTalk 表示持久化后的钉钉通道类型。
	SessionChannelDingTalk = "dingtalk"
	// SessionChannelWeChat 表示持久化后的微信通道类型。
	SessionChannelWeChat = "wechat"
	// SessionChannelWeixinPersonal 表示持久化后的个人微信 iLink 通道类型。
	SessionChannelWeixinPersonal = "weixin-personal"
	// SessionChannelFeishu 表示持久化后的飞书通道类型。
	SessionChannelFeishu = "feishu"

	// AutomationInboxSessionRef 表示定时任务投递到 Agent 时使用的固定内部会话。
	AutomationInboxSessionRef = "automation-inbox"
)

// SessionKeyKind 表示协议族。
type SessionKeyKind string

const (
	// SessionKeyKindAgent 表示 agent 私有运行时。
	SessionKeyKindAgent SessionKeyKind = "agent"
	// SessionKeyKindRoom 表示共享 room 流。
	SessionKeyKindRoom SessionKeyKind = "room"
	// SessionKeyKindUnknown 表示无法识别。
	SessionKeyKindUnknown SessionKeyKind = "unknown"

	roomSharedChatType = "group"
	topicSegment       = "topic"
	accountSegment     = "acct"
)

// SessionKey 表示结构化会话键。
type SessionKey struct {
	Raw            string         `json:"raw"`
	Kind           SessionKeyKind `json:"kind"`
	IsStructured   bool           `json:"is_structured"`
	IsShared       bool           `json:"is_shared"`
	AgentID        string         `json:"agent_id,omitempty"`
	Channel        string         `json:"channel,omitempty"`
	ChatType       string         `json:"chat_type,omitempty"`
	AccountID      string         `json:"account_id,omitempty"`
	Ref            string         `json:"ref,omitempty"`
	ThreadID       string         `json:"thread_id,omitempty"`
	ConversationID string         `json:"conversation_id,omitempty"`
	RoomRef        string         `json:"room_ref,omitempty"`
}

// ErrInvalidSessionKey 表示 session_key 不符合结构化协议。
var ErrInvalidSessionKey = errors.New("invalid structured session_key")

// StructuredSessionKeyError 对齐前端 HTTP 入口的 422 校验错误。
type StructuredSessionKeyError struct {
	Message string
}

func (e StructuredSessionKeyError) Error() string {
	return e.Message
}

func findTopicIndex(parts []string, minIndex int) int {
	for index, value := range parts {
		if value == topicSegment && index >= minIndex {
			return index
		}
	}
	return -1
}

func splitAgentRefParts(parts []string) (string, int, string) {
	if len(parts) > 4 && parts[4] == accountSegment {
		if len(parts) < 7 {
			return "", 0, "session_key must match agent:<agent_id>:<channel>:<chat_type>[:acct:<account_id>]:<ref>[:topic:<thread_id>]"
		}
		accountID := strings.TrimSpace(parts[5])
		if accountID == "" {
			return "", 0, "session_key account_id is required after acct segment"
		}
		return accountID, 6, ""
	}
	return "", 4, ""
}

func agentSessionKeyShapeError() string {
	return "session_key must match agent:<agent_id>:<channel>:<chat_type>[:acct:<account_id>]:<ref>[:topic:<thread_id>]"
}

// GetSessionKeyValidationError 返回结构化 session_key 校验错误。
func GetSessionKeyValidationError(raw string) string {
	normalized := strings.TrimSpace(raw)
	if normalized == "" {
		return "session_key is required"
	}

	if strings.HasPrefix(normalized, string(SessionKeyKindAgent)+":") {
		return validateAgentSessionKey(normalized)
	}

	if strings.HasPrefix(normalized, string(SessionKeyKindRoom)+":") {
		return validateRoomSessionKey(normalized)
	}

	return "session_key must use structured session_key format"
}

func validateAgentSessionKey(sessionKey string) string {
	parts := strings.Split(sessionKey, ":")
	if len(parts) < 5 || strings.TrimSpace(parts[1]) == "" || strings.TrimSpace(parts[2]) == "" || strings.TrimSpace(parts[3]) == "" {
		return agentSessionKeyShapeError()
	}
	_, refStart, splitErr := splitAgentRefParts(parts)
	if splitErr != "" {
		return splitErr
	}
	topicIndex := findTopicIndex(parts, refStart)
	if topicIndex < 0 {
		if strings.TrimSpace(strings.Join(parts[refStart:], ":")) == "" {
			return agentSessionKeyShapeError()
		}
		return ""
	}
	ref := strings.TrimSpace(strings.Join(parts[refStart:topicIndex], ":"))
	threadID := strings.TrimSpace(strings.Join(parts[topicIndex+1:], ":"))
	if ref == "" || threadID == "" {
		return agentSessionKeyShapeError()
	}
	return ""
}

func validateRoomSessionKey(sessionKey string) string {
	parts := strings.Split(sessionKey, ":")
	if len(parts) < 3 || parts[1] != roomSharedChatType || strings.TrimSpace(strings.Join(parts[2:], ":")) == "" {
		return "session_key must match room:group:<conversation_id>"
	}
	return ""
}

// RequireStructuredSessionKey 要求必须是结构化 session_key。
func RequireStructuredSessionKey(raw string) (string, error) {
	if message := GetSessionKeyValidationError(raw); message != "" {
		return "", StructuredSessionKeyError{Message: message}
	}
	return strings.TrimSpace(raw), nil
}

// ParseSessionKey 解析 session_key。
func ParseSessionKey(raw string) SessionKey {
	normalized := strings.TrimSpace(raw)
	validationError := GetSessionKeyValidationError(normalized)
	result := SessionKey{
		Raw:          normalized,
		Kind:         SessionKeyKindUnknown,
		IsStructured: validationError == "",
	}

	if strings.HasPrefix(normalized, string(SessionKeyKindAgent)+":") {
		parts := strings.Split(normalized, ":")
		result.Kind = SessionKeyKindAgent
		if len(parts) > 1 {
			result.AgentID = strings.TrimSpace(parts[1])
		}
		if len(parts) > 2 {
			result.Channel = strings.TrimSpace(parts[2])
		}
		result.ChatType = "dm"
		if len(parts) > 3 {
			if chatType := strings.TrimSpace(parts[3]); chatType != "" {
				result.ChatType = chatType
			}
		}

		// `:topic:` 是保留边界，ref 允许带冒号，但不能跨过这个边界。
		accountID, refStart, splitErr := splitAgentRefParts(parts)
		if splitErr != "" {
			return result
		}
		result.AccountID = accountID
		topicIndex := findTopicIndex(parts, refStart)
		if topicIndex >= 0 {
			result.Ref = strings.TrimSpace(strings.Join(parts[refStart:topicIndex], ":"))
			result.ThreadID = strings.TrimSpace(strings.Join(parts[topicIndex+1:], ":"))
			return result
		}

		if len(parts) > refStart {
			result.Ref = strings.TrimSpace(strings.Join(parts[refStart:], ":"))
		}
		return result
	}

	if strings.HasPrefix(normalized, string(SessionKeyKindRoom)+":") {
		parts := strings.Split(normalized, ":")
		conversationID := ""
		if len(parts) > 2 {
			conversationID = strings.TrimSpace(strings.Join(parts[2:], ":"))
		}
		result.Kind = SessionKeyKindRoom
		result.IsShared = validationError == ""
		result.ChatType = roomSharedChatType
		if len(parts) > 1 {
			if chatType := strings.TrimSpace(parts[1]); chatType != "" {
				result.ChatType = chatType
			}
		}
		result.Ref = conversationID
		result.RoomRef = conversationID
		result.ConversationID = conversationID
		return result
	}

	return result
}

// BuildAgentSessionKey 构建 agent 作用域 key。
func BuildAgentSessionKey(agentID string, channel string, chatType string, ref string, threadID string) string {
	return BuildAgentAccountSessionKey(agentID, channel, chatType, "", ref, threadID)
}

// BuildAgentAccountSessionKey 构建带外部通道账号作用域的 agent key。
func BuildAgentAccountSessionKey(agentID string, channel string, chatType string, accountID string, ref string, threadID string) string {
	accountID = strings.TrimSpace(accountID)
	ref = strings.TrimSpace(ref)
	threadID = strings.TrimSpace(threadID)
	base := fmt.Sprintf(
		"agent:%s:%s:%s:%s",
		strings.TrimSpace(agentID),
		NormalizeSessionKeyChannelSegment(channel),
		NormalizeSessionChatType(chatType),
		ref,
	)
	if accountID != "" {
		base = fmt.Sprintf(
			"agent:%s:%s:%s:%s:%s:%s",
			strings.TrimSpace(agentID),
			NormalizeSessionKeyChannelSegment(channel),
			NormalizeSessionChatType(chatType),
			accountSegment,
			accountID,
			ref,
		)
	}
	if threadID == "" {
		return base
	}
	return base + ":" + topicSegment + ":" + threadID
}

// BuildRoomSharedSessionKey 构建共享 room key。
func BuildRoomSharedSessionKey(conversationID string) string {
	return "room:" + roomSharedChatType + ":" + strings.TrimSpace(conversationID)
}

// BuildRoomAgentSessionKey 构建 Room 成员侧的 agent session_key。
func BuildRoomAgentSessionKey(conversationID string, agentID string, roomType string) string {
	chatType := "group"
	if strings.TrimSpace(roomType) == "dm" {
		chatType = "dm"
	}
	return BuildAgentSessionKey(strings.TrimSpace(agentID), SessionChannelWebSocketSegment, chatType, strings.TrimSpace(conversationID), "")
}

// IsRoomSharedSessionKey 判断是否为 Room 共享消息流 key。
func IsRoomSharedSessionKey(raw string) bool {
	parsed := ParseSessionKey(raw)
	return parsed.Kind == SessionKeyKindRoom && parsed.IsStructured && parsed.ConversationID != ""
}

// ParseRoomConversationID 读取 Room 共享流里的 conversation_id。
func ParseRoomConversationID(raw string) string {
	parsed := ParseSessionKey(raw)
	if parsed.Kind != SessionKeyKindRoom {
		return ""
	}
	return parsed.ConversationID
}

// NormalizeSessionKeyChannelSegment 把外部输入统一成 session_key 使用的 channel 段。
func NormalizeSessionKeyChannelSegment(channel string) string {
	switch strings.ToLower(strings.TrimSpace(channel)) {
	case SessionChannelWebSocketSegment, SessionChannelWebSocket:
		return SessionChannelWebSocketSegment
	case SessionChannelDiscordSegment, SessionChannelDiscord:
		return SessionChannelDiscordSegment
	case SessionChannelTelegramSegment, SessionChannelTelegram:
		return SessionChannelTelegramSegment
	case SessionChannelDingTalkSegment, SessionChannelDingTalk:
		return SessionChannelDingTalkSegment
	case SessionChannelWeChatSegment, SessionChannelWeChat:
		return SessionChannelWeChatSegment
	case SessionChannelWeixinPersonalSegment:
		return SessionChannelWeixinPersonalSegment
	case SessionChannelFeishuSegment, SessionChannelFeishu:
		return SessionChannelFeishuSegment
	case SessionChannelInternalSegment:
		return SessionChannelInternalSegment
	default:
		return strings.TrimSpace(channel)
	}
}

// NormalizeStoredChannelType 把 channel 归一成持久化和运行时使用的名称。
func NormalizeStoredChannelType(channel string) string {
	switch strings.ToLower(strings.TrimSpace(channel)) {
	case SessionChannelWebSocketSegment, SessionChannelWebSocket:
		return SessionChannelWebSocket
	case SessionChannelDiscordSegment, SessionChannelDiscord:
		return SessionChannelDiscord
	case SessionChannelTelegramSegment, SessionChannelTelegram:
		return SessionChannelTelegram
	case SessionChannelDingTalkSegment, SessionChannelDingTalk:
		return SessionChannelDingTalk
	case SessionChannelWeChatSegment, SessionChannelWeChat:
		return SessionChannelWeChat
	case SessionChannelWeixinPersonalSegment:
		return SessionChannelWeixinPersonal
	case SessionChannelFeishuSegment, SessionChannelFeishu:
		return SessionChannelFeishu
	case SessionChannelInternalSegment:
		return SessionChannelInternalSegment
	default:
		return strings.TrimSpace(channel)
	}
}

// NormalizeSessionChatType 统一 chat_type 的默认值和枚举。
func NormalizeSessionChatType(chatType string) string {
	switch strings.ToLower(strings.TrimSpace(chatType)) {
	case "", "dm":
		return "dm"
	case "group":
		return "group"
	default:
		return strings.TrimSpace(chatType)
	}
}
